contributors/design-proposals/storage/postpone-pvc-deletion-if-used-in-a-pod.md

```diff
@@ -0,0 +1,101 @@
+
+Status: Proposal
+
+Version: GA
+
+Implementation Owner: @pospispa
+
+## Motivation
+
+User can delete a Persistent Volume Claim (PVC) that is being used by a pod. This may have negative impact on the pod and it may result in data loss.
+
+For more details see issue https://github.com/kubernetes/kubernetes/issues/45143
+
+## Proposal
+
+Postpone the PVC deletion until the PVC is not used by any pod.
+
+## User Experience
+
+### Use Cases
+
+1. User deletes a PVC that is being used by a pod. This may have negative impact on the pod and may result in data loss. As a user, I want that any PVC deletion does not have any negative impact on any pod. As a user, I do not want to experience data loss.
+
+#### Scenarios for data loss
+Depending on the storage type the data loss occurs in one of the below scenarios:
+- in case the dynamic provisioning is used and reclaim policy is `Delete` the PVC deletion triggers deletion of the associated storage asset and PV.
+- the same as above applies for the static provisioning and `Delete` reclaim policy.
+
+## Implementation
+
+### API Sever, PVC Create
+When a PVC is created API server will automatically add finalizer into the PVC's metadata.
+
+### Scheduler
+Scheduler will check if a pod uses a PVC and if any of the PVCs has `deletionTimestamp` set. In case this is true the scheduling of the pod will end up with an error: "PVC (%pvcName) is in scheduled for deletion state".
+
+### PVC Deletion Controller
+PVC deletion controller is a new internal controller.
+
+PVC deletion controller can list pods and watch for pods. It caches status of pods that use a PVC. It also watches for PVC deletes.
+
+PVC deletion controller reacts to PVC delete in one of the below ways:
+- In case the PVC is not used by a pod: removes the finalizer information from the PVC metadata.
+- In case the PVC is used by a pod: adds the PVC into a set of PVCs that are in use and are waiting for deletion.
+
+PVC deletion controller reacts to pod termination in the below way:
+- In case the pod was using PVC(s) that are waiting for deletion and these PVC(s) are not used by any other pods: removes the finalizer information from the PVC(s) metadata.
+
+Note: the PVC deletion controller may be down for some time.
+
+PVC deletion controller after it starts gets current status of pods and PVCs from API server. Afterwards, it does the below:
+- Removes the finalizer information from the PVC metadata for all PVC that have `deletionTimestamp` set and are not used by a pod.
+- Adds all PVCs that have the `deletionTimestamp` set and are used by a pod into a set of PVCs that are in use and are waiting for deletion.
+
+### Known Race Condition
+1. Pod A is currently using PVC.
+2. Pod A is deleted.
+3. Soon after, Pod B is created and scheduled that uses PVC.
+4. PVC is deleted.
+5. PVC deletion controller gets Pod delete for Pod A, sees no other Pod is using the PVC in its cache, and removes finalizer.
```

Comment:
Could the race be solved if instead of removing the finalizer immediately, you resync all the Pods first?

Reply:
If PVC deletion controller resyncs all the pods before removing the finalize is not enough. There must be a PVC resync done in scheduler or in kubelet. IMHO, these two resyncs will solve the race. I'll update the proposal accordingly.

Reply:
> There must be a PVC resync done in scheduler or in kubelet. IMHO, these two resyncs will solve the race. I'll update the proposal accordingly.

Not sure what you mean by resync... the kubelet does a live lookup of the PVC when setting up the pod's volume.

Reply:
> Not sure what you mean by resync in the kubelet... the kubelet does a live lookup of the PVC when setting up the pod's volume. If the PVC has a deletionTimestamp, the kubelet should not continue setting up the volume.

I didn't know that kubelet does a live lookup of the PVC(s) used by the pod so I wanted to add it.

If the PVC has a deletionTimestamp kubelet must not start the pod.